[install CentOS] httpd.conf configuration instructions (#5533)
aces/Loris: README.CentOS7.md

 * DocumentRoot should point to `/var/www/loris/htdocs`
 * The `smarty/templates_c/` directory must be writable by Apache (e.g. by running: `sudo chgrp -R httpd /var/www/loris/smarty/templates_c` and `sudo chmod 775 /var/www/loris/smarty/templates_c`).

> Create the Apache configuration `/etc/httpd/conf.d/loris.conf` for your LORIS environment. You can find an example in `docs/config/apache2-site` for setup of `VirtualHost` in the loris.conf file you will create. Adjust the parameters according to your configuration.

Reviewer: You lost the part about the adjustments to be made:

```
- %LORISROOT%    i.e. /var/www/loris
- %PROJECTNAME%  i.e loris
- %LOGDIRECTORY%  .i.e /var/log/httpd
```

Author: @driusan added again.

 * DocumentRoot should point to `/var/www/loris/htdocs`
 * The `smarty/templates_c/` directory must be writable by Apache (e.g. by running: `sudo chgrp -R httpd /var/www/loris/smarty/templates_c` and `sudo chmod 775 /var/www/loris/smarty/templates_c`).

> Create the Apache configuration `/etc/httpd/conf.d/loris.conf` for your LORIS environment. You can find an example in `loris/docs/config/apache2-site` for setup of `VirtualHost` in the loris.conf file you will create. Adjust the parameters according to your configuration.

Reviewer 2: ```suggestion
Create the Apache configuration `/etc/httpd/conf.d/loris.conf` for your LORIS environment. You can find an example in `loris/docs/config/apache2-site` for setup of `<VirtualHost>` in the loris.conf file you will create. Adjust the parameters according to your configuration.
```

 * DocumentRoot should point to `/var/www/loris/htdocs`
 * The `smarty/templates_c/` directory must be writable by Apache (e.g. by running: `sudo chgrp -R httpd /var/www/loris/smarty/templates_c` and `sudo chmod 775 /var/www/loris/smarty/templates_c`).

Modify the Apache configuration file for our production environment.
```

> sudo open -e /usr/local/etc/httpd/httpd.conf

Reviewer 2: this OS x tools isn't the same on Linux.

People can use vi (installed by default), nano (quite often install by default now), emacs, ...
Would be good to either 
  - tell user to use their preferred file editor or
  - give them instruction on how to navigate, change detail, save, ...

Plus, the http config file is `/etc/httpd/conf/httpd.conf`

Author: @PapillonMcGill thank you. Please see new commit for change.

Reviewer 2: I don't think that target user will no how to use `vi`

> ```

Reviewer: This is a sentence, it shouldn't be in a codeblock.

**a.** Find & modify the `ServerName` to:
```

> ServerName your.Loris.url.here:443

Reviewer: This document doesn't have any instruction on how to setup SSL that I see, so this port is wrong.

Author: @driusan I changed it to 80 in latest commit.

Reviewer: there's no reason to include a port number at all in the server name.

Reviewer 3: I think it's fine to prompt `(typically port 443 for SSL)`

> ServerName <production_server>

Reviewer 2: not wise to use <> to indicate choice as it is used in tag.
A better choice would be something like `your.Loris.url.here:443`.   and port number 80 or 443

should be done at <VirtualHost> tag as done in Ubuntu by default loris install

Author: @PapillonMcGill thank you. Please see new commit for change.

Reviewer 2: What if user do not use /var/www/loris as a base dir?
Still not showing as close as possible the specification in <baseLorisDir>/docs/config/apache2-site.

Reviewer 3: There should be other examples in our docs of the `$yourURL` and  `$lorisroot` or `$basePath` type placeholder (i forget which) -- consistency with those is ideal.

 * DocumentRoot should point to `/var/www/loris/htdocs`
 * The `smarty/templates_c/` directory must be writable by Apache (e.g. by running: `sudo chgrp -R httpd /var/www/loris/smarty/templates_c` and `sudo chmod 775 /var/www/loris/smarty/templates_c`).

> Create the Apache configuration `/etc/httpd/conf.id/loris.conf` for our production environment.

Reviewer 2: ```suggestion
Create the Apache configuration `/etc/httpd/conf.d/loris.conf` for your LORIS environment. Adjust the following parameters to according to your configuration

- %LORISROOT%    i.e. /var/www/loris
- %PROJECTNAME%  i.e loris
- %LOGDIRECTORY%  .i.e /var/log/httpd
```
could be setting up a staging or dev instance!

example for placeholder could be ignored.

```
<VirtualHost *:80>  # change from 80 to 443 if you enable SSL
        ServerAdmin webmaster@localhost

>         ServerName your.Loris.url.here

Reviewer 2: ```suggestion
```
To follow documentations in code sections and install.sh

>         DocumentRoot /var/www/loris/htdocs/

Reviewer 2: ```suggestion
        DocumentRoot /var/www/%LORISROOT%/htdocs/
```

        <Directory />
                Options FollowSymLinks
                AllowOverride All
        </Directory>

>         <Directory /var/www/loris/htdocs/>

Reviewer 2: ```suggestion
        <Directory /var/www/%LORISROOT%/htdocs/>
```

                Options Indexes FollowSymLinks MultiViews
                AllowOverride All
                Order allow,deny
                allow from all
        </Directory>

>         php_value include_path .:/usr/share/php:/var/www/loris/project/libr$

Reviewer 2: ```suggestion
         php_value include_path .:/usr/share/php:%LORISROOT%/project/libraries:%LORISROOT%/php/libraries
```

        #DirectoryIndex main.php index.html

        ErrorLog /var/log/httpd/loris-error.log

        # Possible values include: debug, info, notice, warn, error, crit,
        # alert, emerg.
        LogLevel warn

>         CustomLog /var/log/httpd/loris-access.log combined

Reviewer 2: ```suggestion
        CustomLog %LOGDIRECTORY%/loris-access.log combined
```

        ServerSignature Off

        #SSLEngine Off  # change to On to enable, after updating cert paths$
        #SSLCertificateFile /etc/apache2/ssl/loris-cert.pem
        #SSLCertificateKeyFile /etc/apache2/ssl/loris-key.pem

>         #SSLCACertificateFile /etc/apache2/ssl/CA-cacert.pem

Reviewer 2: Ubuntu path!
```
#SSLCertificateFile /etc/pki/tls/certs/%PROJECTNAME%.pem
#SSLCertificateKeyFile /etc/pki/tls/private/%PROJECTNAME%.key
#SSLCACertificateFile /etc/pki/tls/certs/CA-cert.pem
```

 * DocumentRoot should point to `/var/www/loris/htdocs`
 * The `smarty/templates_c/` directory must be writable by Apache (e.g. by running: `sudo chgrp -R httpd /var/www/loris/smarty/templates_c` and `sudo chmod 775 /var/www/loris/smarty/templates_c`).

Create the Apache configuration `/etc/httpd/conf.d/loris.conf` for your LORIS environment. Adjust the following parameters to according to your configuration

- %LORISROOT%    i.e. /var/www/loris
- %PROJECTNAME%  i.e loris
- %LOGDIRECTORY%  .i.e /var/log/httpd
```

> <VirtualHost *:80>  # change from 80 to 443 if you enable SSL

Reviewer: 1. Why is this duplicating the docs/config/apache2-site file instead of simply referencing it? It's not going to get maintained here
2. the paths shouldn't include "/var/www", that's part of the %LORISROOT% variable.

Author: @driusan can you review the latest changes. I referenced the loris/docs/config/apache2-site and removed the VirtualHost example.

Modify the Apache configuration file for our production environment.
```

> sudo open -e /etc/httpd/conf/httpd.conf

Reviewer 2: Open is not the same thing on Linux as on OS x.

Author: @PapillonMcGill do you know what command is typically used for CentOS to open a file for a user to edit. Most guides online use vi and I'm not sure if nano comes with all CentOS installations. I think it does but maybe user doesn't know how to use it as well.

Reviewer 2: Vi is always there, nano is also on a VM I have and I haven't specifically installed it. So it's probably there by default now on CentOS.Emacs is not by default.
I will check tonight with 2 machines running CentOS 7 on which I know I haven't installed nano.

Reviewer: ed is the standard text editor.

Reviewer 2: @maltheism Would you provide instruction on how to use `ed`? I beth not many target LORIS user know how to use ed. I have to say that I forgot after all those years.

Author: @PapillonMcGill @driusan I'm looking at ed and for me it seems more confusing then vi

Reviewer 3: I don't think we should specify a text editor.  would suggest something more general like: 
`Edit  /etc/httpd/conf/httpd.conf as follows (may require sudo-ing the text editor command)`

Reviewer: ?

Reviewer 2: Just configured a minimalist CentOS7 vm, ed, nano and vi are installed. You have the choice of editor.

Author: @PapillonMcGill @driusan I added the comment from @christinerogers and maybe we should just leave the editor choice to the user.

Reviewer 2: @maltheism Could you write the instruction with <VirtualHost> instead of serverName and follow the instruction in https://github.com/aces/Loris/blob/major/docs/config/apache2-site.
Putting a thumb-up twice (see higher) when requested and no doing it doesn't help.

Author: @PapillonMcGill I'm a little confused by how you wrote your comment. Do you want me to link to ../config/apache2-site inside the README.CentOS7.md document? The apache2-site document doesn't solve the requirement of having a ServerName set in the /etc/httpd/conf/httpd.conf from what I've experienced. Also running the tools/install.sh script makes the apache2-site configuration but doesn't solve setting up the /etc/httpd/conf/httpd.conf

Reviewer 2: No, I am not asking to link the document but to re-use the setting from that documents. 
Using the 'ServerName' directive is restrictive for Apache. Now, the practice is to use <VirtualHost> directive. 
I am not asking to modify the install script either as it only do the change on Ubuntu system.  Simply to have a similar configuration on both CentOS and Ubuntu. Apache work the same on both platform even if both distros use different config files.

Author: @PapillonMcGill okay I think I understand what you're implying. Do you want me to remove the instructions to modify `ServerName` because the `VirtualHost` file should render that task unnecessary? I didn't have success with the `VirtualHost` that's made with the install.sh working when going through the installation process of CentOS for Loris. I had to setup the `ServerName` before Loris would load in the browser.

Reviewer 2: With `<VirtualHost>`, you can't use `ServerName`. They are mutually exclusive.
Here is an example from an existing server (with some obfuscation)
```
<VirtualHost *:80>  
        ServerName example.loris.ca
	ServerAdmin webmaster@localhost
	
	DocumentRoot /var/www/Loris/htdocs
	<Directory />
		Options FollowSymLinks
		AllowOverride All
	</Directory>
	<Directory /var/www/Loris/>
		Options Indexes FollowSymLinks MultiViews
		AllowOverride All
		Order allow,deny
		allow from all
	</Directory>

	php_value include_path .:/usr/share/php:/var/www/Loris/project/libraries:/var/www/Loris/php/libraries

	DirectoryIndex main.php index.html

	LogFormat "%{X-Forwarded-For}i %l %u %t \"%r\" %>s %b \"%{Referer}i\" \"%{User-Agent}i\"" combined
	ErrorLog /var/log/httpd/loris-error.log

	# Possible values include: debug, info, notice, warn, error, crit,
	# alert, emerg.
	LogLevel warn 

	CustomLog /var/log/httpd/loris-access.log combined
	ServerSignature Off

</VirtualHost>
```

Author: @PapillonMcGill okay, do I add instructions telling the user to set `ServerName` to be equivalent to what `VirtualHost` is configured to be? Also that VirtualHost is made with the tools/install.sh script.

Reviewer 2: No, there must not be any `ServerName` directive. You may comment any line starting with `ServerName` or delete them.

Author: @PapillonMcGill okay, I'm going to remove the modify httpd.conf instructions and go through the installation process again to see if ServerName is uncommented or not.

Author: Hi @PapillonMcGill, I just went through the installation process on a new CentOS 7 vm.

Here is my /etc/httpd/conf.d/loris.conf
```
<VirtualHost *:80>  # change from 80 to 443 if you enable SSL
        ServerAdmin webmaster@localhost
        ServerName localhost

        DocumentRoot /var/www/loris/htdocs/
        <Directory />
                Options FollowSymLinks
                AllowOverride All
        </Directory>
        <Directory /var/www/loris/htdocs/>
                Options Indexes FollowSymLinks MultiViews
                AllowOverride All
                Order allow,deny
                allow from all
        </Directory>

        php_value include_path .:/usr/share/php:/var/www/loris/project/libr$

        #DirectoryIndex main.php index.html

        ErrorLog /var/log/httpd/loris-error.log

        # Possible values include: debug, info, notice, warn, error, crit,
        # alert, emerg.
        LogLevel warn

        CustomLog /var/log/httpd/loris-access.log combined
        ServerSignature Off

        #SSLEngine Off  # change to On to enable, after updating cert paths$
        #SSLCertificateFile /etc/apache2/ssl/loris-cert.pem
        #SSLCertificateKeyFile /etc/apache2/ssl/loris-key.pem
        #SSLCACertificateFile /etc/apache2/ssl/CA-cacert.pem

</VirtualHost>
```

and my /etc/httpd/conf/httpd.conf
```
#
# This is the main Apache HTTP server configuration file.  It contains the
# configuration directives that give the server its instructions.
# See <URL:http://httpd.apache.org/docs/2.4/> for detailed information.
# In particular, see
# <URL:http://httpd.apache.org/docs/2.4/mod/directives.html>
# for a discussion of each configuration directive.
#
# Do NOT simply read the instructions in here without understanding
# what they do.  They're here only as hints or reminders.  If you are unsure
# consult the online docs. You have been warned.
#
# Configuration and logfile names: If the filenames you specify for many
# of the server's control files begin with "/" (or "drive:/" for Win32), the
# server will use that explicit path.  If the filenames do *not* begin
# with "/", the value of ServerRoot is prepended -- so 'log/access_log'
# with ServerRoot set to '/www' will be interpreted by the
# server as '/www/log/access_log', where as '/log/access_log' will be
# interpreted as '/log/access_log'.

#
# ServerRoot: The top of the directory tree under which the server's
# configuration, error, and log files are kept.
#
# Do not add a slash at the end of the directory path.  If you point
# ServerRoot at a non-local disk, be sure to specify a local disk on the
# Mutex directive, if file-based mutexes are used.  If you wish to share the
# same ServerRoot for multiple httpd daemons, you will need to change at
# least PidFile.
#
ServerRoot "/etc/httpd"

#
# Listen: Allows you to bind Apache to specific IP addresses and/or
# ports, instead of the default. See also the <VirtualHost>
# directive.
#
# Change this to Listen on specific IP addresses as shown below to
# prevent Apache from glomming onto all bound IP addresses.
#
#Listen 12.34.56.78:80
Listen 80

#
# Dynamic Shared Object (DSO) Support
#
# To be able to use the functionality of a module which was built as a DSO $
# have to place corresponding `LoadModule' lines at this location so the
# directives contained in it are actually available _before_ they are used.
# Statically compiled modules (those listed by `httpd -l') do not need
# to be loaded here.
#
# Example:
# LoadModule foo_module modules/mod_foo.so
#
Include conf.modules.d/*.conf

#
# If you wish httpd to run as a different user or group, you must run
# httpd as root initially and it will switch.
#
# User/Group: The name (or #number) of the user/group to run httpd as.
# It is usually good practice to create a dedicated user and group for
# running httpd, as with most system services.
#
User apache
Group apache

# 'Main' server configuration
#
# The directives in this section set up the values used by the 'main'
# server, which responds to any requests that aren't handled by a
# <VirtualHost> definition.  These values also provide defaults for
# any <VirtualHost> containers you may define later in the file.
#
# All of these directives may appear inside <VirtualHost> containers,
# in which case these default settings will be overridden for the
# virtual host being defined.
#

#
# ServerAdmin: Your address, where problems with the server should be
# e-mailed.  This address appears on some server-generated pages, such
# as error documents.  e.g. admin@your-domain.com
#
ServerAdmin root@localhost

#
# ServerName gives the name and port that the server uses to identify itsel$
# This can often be determined automatically, but we recommend you specify
# it explicitly to prevent problems during startup.
#
# If your host doesn't have a registered DNS name, enter its IP address her$
#
#ServerName www.example.com:80

#
# Deny access to the entirety of your server's filesystem. You must
# explicitly permit access to web content directories in other
# <Directory> blocks below.
#
<Directory />
    AllowOverride none
    Require all denied
</Directory>

#
# Note that from this point forward you must specifically allow
# particular features to be enabled - so if something's not working as
# you might expect, make sure that you have specifically enabled it
# below.
#

#
# DocumentRoot: The directory out of which you will serve your
# documents. By default, all requests are taken from this directory, but
# symbolic links and aliases may be used to point to other locations.
#
DocumentRoot "/var/www/html"

#
# Relax access to content within /var/www.
#
<Directory "/var/www">
    AllowOverride None
    # Allow open access:
    Require all granted
</Directory>

# Further relax access to the default document root:
<Directory "/var/www/html">
    #
    # Possible values for the Options directive are "None", "All",
    # or any combination of:
    #   Indexes Includes FollowSymLinks SymLinksifOwnerMatch ExecCGI MultiV$
    #
    # Note that "MultiViews" must be named *explicitly* --- "Options All"
    # doesn't give it to you.
    #
    # The Options directive is both complicated and important.  Please see
    # http://httpd.apache.org/docs/2.4/mod/core.html#options
    # for more information.
#
    Options Indexes FollowSymLinks

    #
    # AllowOverride controls what directives may be placed in .htaccess fil$
    # It can be "All", "None", or any combination of the keywords:
    #   Options FileInfo AuthConfig Limit
    #
    AllowOverride None

    #
    # Controls who can get stuff from this server.
    #
    Require all granted
</Directory>

#
# DirectoryIndex: sets the file that Apache will serve if a directory
# is requested.
#
<IfModule dir_module>
    DirectoryIndex index.html
</IfModule>

#
# The following lines prevent .htaccess and .htpasswd files from being
# viewed by Web clients.
#
<Files ".ht*">
    Require all denied
</Files>

#
# ErrorLog: The location of the error log file.
# If you do not specify an ErrorLog directive within a <VirtualHost>
# container, error messages relating to that virtual host will be
# logged here.  If you *do* define an error logfile for a <VirtualHost>
# container, that host's errors will be logged there and not here.
#
ErrorLog "logs/error_log"

#
# LogLevel: Control the number of messages logged to the error_log.
# Possible values include: debug, info, notice, warn, error, crit,
# alert, emerg.
#
LogLevel warn

```

I'm getting a blank page in my browser and I don't think the `VirtualHost` is working. Any suggestions?

Reviewer 2: Have you set the value for the following parameter in Config table?

    base
    url
    host

Reviewer 2: check value for `php_value include_path` in loris.conf` and also, remove line about `ssl`, these are for Ubuntu, not CentOS.

Reviewer 2: some error in my config, it should start with
```
<VirtualHost example.loris.ca:80>  
	ServerAdmin webmaster@localhost
	
	DocumentRoot /var/www/Loris/htdocs
...
```

Author: @PapillonMcGill thanks, turns out the loris-error.log was showing PDO was not found.
So I needed to do `sudo yum --enablerepo=remi-php72 install php-pdo php-pdo_mysql` and now it works.
